### PR TITLE
small features for reindex and updatemapping script

### PR DESCRIPTION
**Pull request by NickPapacostas**
reindex script takes optional "from date", update mappings takes optional index name

### CODE REVIEW

File: scripts/src/main/scala/com/gu/mediaservice/scripts/EsScript.scala [hunk 1]
```diff
@@ -99,13 +104,22 @@ object Reindex extends EsScript {
...
+          val recentChangesQuery = rangeQuery("lastModified").from(fromTime.get).to(DateTime.now)
+          reindexScroll(query.setQuery(recentChangesQuery).execute.actionGet)
+        }
 
         // TODO: deleteIndex when we are confident
         client.close()
       }
     }
-    EsClient.reindex
+    if(extraArgs.isEmpty) {
```
Comment (kenoir) [inline scripts/src/main/scala/com/gu/mediaservice/scripts/EsScript.scala]: I'd avoid `if` statements like this wherever possible and either `foldLeft` or return a value:

```scala
val date = if(extraArgs.isEmpty) None else Some((DateTime.parse(extraArgs(0))))
EsClient.reindex(date)
```

or

```scala
EsClient.reindex(
  extraArgs.foldLeft(None)(Some(DateTime.parse(extraArgs(0))))
)
```

Comment (kenoir) [inline scripts/src/main/scala/com/gu/mediaservice/scripts/EsScript.scala]: Goes for the `if` statements elsewhere too.

Comment (LATaylor-guardian) [inline scripts/src/main/scala/com/gu/mediaservice/scripts/EsScript.scala]: A statement in the form:

val date = if(extraArgs.isEmpty) None else Some((DateTime.parse(extraArgs(0))))
EsClient.reindex(date)

is certainly more accessible to non-native Scala developers or newcomers than foldLeft.

Comment (jamesgorrie) [inline scripts/src/main/scala/com/gu/mediaservice/scripts/EsScript.scala]: But it's quite a good thing to find out about if you don't know it.

**Comment by NickPapacostas** [comment]
tried to find a happy medium :) foldLeft didn't make it in though :disappointed:

File: scripts/src/main/scala/com/gu/mediaservice/scripts/EsScript.scala [hunk 2]
```diff
@@ -2,12 +2,19 @@ package com.gu.mediaservice.scripts
 
 import org.elasticsearch.action.admin.indices.close.CloseIndexRequest
 import org.elasticsearch.action.admin.indices.open.OpenIndexRequest
-import org.elasticsearch.action.search.SearchResponse
-import org.elasticsearch.search.sort.SortOrder
-import org.elasticsearch.index.query.QueryBuilders.matchAllQuery
+import org.elasticsearch.action.bulk.{BulkRequestBuilder}
```
Comment (akash1810) [inline scripts/src/main/scala/com/gu/mediaservice/scripts/EsScript.scala]: don't need braces with singular imports

File: scripts/src/main/scala/com/gu/mediaservice/scripts/EsScript.scala [hunk 3]
```diff
@@ -42,81 +49,132 @@ object MoveIndex extends EsScript {
...
+    val newIndex = EsClient.nextIndex
+
+    val from = if(args.isEmpty) None else Some(DateTime.parse(args(0)))
+    validateCurrentState(EsClient, from)
+    Await.result(reindex(from, EsClient), Duration.Inf)
+
+    def reindex(from: Option[DateTime], esClient: ElasticSearchClient) : Future[SearchResponse] = {
+      def _scroll(scroll: SearchResponse, done: Long = 0): Future[SearchResponse] = {
+        val client = esClient.client
+        val doing = done + scrollSize
```
Comment (akash1810) [inline scripts/src/main/scala/com/gu/mediaservice/scripts/EsScript.scala]: interesting naming :stuck_out_tongue:

File: scripts/src/main/scala/com/gu/mediaservice/scripts/EsScript.scala [hunk 4]
```diff
@@ -42,81 +49,132 @@ object MoveIndex extends EsScript {
...
+        }
+      }
 
-            bulk.execute.actionGet
-            reindexScroll(client.prepareSearchScroll(scroll.getScrollId)
-              .setScroll(scrollTime).execute.actionGet, doing)
-          } else {
+      def query(from: Option[DateTime]) : SearchRequestBuilder = {
+        val queryType = from.map(time =>
+          rangeQuery("lastModified").from(from.get).to(DateTime.now)
```
Comment (akash1810) [inline scripts/src/main/scala/com/gu/mediaservice/scripts/EsScript.scala]: minor: might want to validate from is always less than `DateTime.now` as its an parameter

File: scripts/src/main/scala/com/gu/mediaservice/scripts/EsScript.scala [hunk 5]
```diff
@@ -42,81 +49,132 @@ object MoveIndex extends EsScript {
...
+          .setTypes(imageType)
+          .setScroll(scrollTime)
+          .setSize(scrollSize)
+          .setQuery(queryType)
+      }
 
-        Thread.sleep(1000)
 
-        reindexScroll(query.execute.actionGet)
+      if(from.isEmpty) {
```
Comment (akash1810) [inline scripts/src/main/scala/com/gu/mediaservice/scripts/EsScript.scala]: its probably obvious, but I don't quite understand this `if` - can you help me?

Comment (NickPapacostas) [inline scripts/src/main/scala/com/gu/mediaservice/scripts/EsScript.scala]: it's not clear! maybe i should throw it in a nicely named method. 

If the user supplies a "from" time (or it's been called recursively) then don't create a new index to populate. Basically, if you're asking for a partial update, then you don't want a new index.

File: scripts/src/main/scala/com/gu/mediaservice/scripts/EsScript.scala (hunk 4 above)
Comment (NickPapacostas) [inline scripts/src/main/scala/com/gu/mediaservice/scripts/EsScript.scala]: good call!

File: scripts/src/main/scala/com/gu/mediaservice/scripts/EsScript.scala (hunk 3 above)
Comment (NickPapacostas) [inline scripts/src/main/scala/com/gu/mediaservice/scripts/EsScript.scala]: haha i'm not sure if you're referring to "doing" but i borrowed it from the old code... funny how things dont seem funny after looking at them for a month >.< how about "currentPage" or "currentBatch"

File: scripts/src/main/scala/com/gu/mediaservice/scripts/EsScript.scala (hunk 5 above)
Comment (akash1810) [inline scripts/src/main/scala/com/gu/mediaservice/scripts/EsScript.scala]: Ah! That makes sense. (could you add a comment for my simile future self? :smile:)

File: scripts/src/main/scala/com/gu/mediaservice/scripts/EsScript.scala [hunk 6]
```diff
@@ -42,81 +49,134 @@ object MoveIndex extends EsScript {
...
+      if(esClient.getCurrentIndices.isEmpty) raise("no index with the 'write' alias exists")
+      if((esClient.getCurrentIndices.length == 2) && from.isEmpty)
+        raise("there are two indices with 'write' alias, check your properties file or use http://localhost:9200/_plugin/head/ ")
+      if(from.exists(_.isAfter(DateTime.now())))
+        raise("DateTime parameter 'from' must be earlier than the current time" )
+    }
 
-        def reindexScroll(scroll: SearchResponse, done: Long = 0) {
+    val imageType = "image"
+    val scrollTime = new TimeValue(5 * 60 * 1000) // 5 minutes in milliseconds
```
Comment (kenoir) [inline scripts/src/main/scala/com/gu/mediaservice/scripts/EsScript.scala]: You might well be able to express this in code using the Joda lib, (http://joda-time.sourceforge.net/apidocs/org/joda/time/Period.html) but it's probably not worth it.

**Comment by kenoir** [comment]
:+1:

**Pull request merged**